File: src/dungeonbreak_narrative/escape_the_dungeon/narrative/deeds.py
"""DeeD models and canonical text vectorization."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable

from .projection import EmbeddingRecord, EmbeddingStore, normalize_text


def _clean_tags(values: Iterable[str]) -> tuple[str, ...]:
    rows = {normalize_text(value) for value in values if normalize_text(value)}
    return tuple(sorted(rows))


@dataclass(frozen=True)
class Deed:
    deed_id: str
    actor_id: str
    actor_name: str
    deed_type: str
    title: str
    summary: str
    depth: int
    chapter_number: int
    room_id: str
    room_feature: str
    turn_index: int
    tags: tuple[str, ...] = ()
    outcome_tags: tuple[str, ...] = ()
    tone_tags: tuple[str, ...] = ()

    def canonical_text(self) -> str:
        tags_text = ", ".join(_clean_tags(self.tags))
        outcome_text = ", ".join(_clean_tags(self.outcome_tags))
        tone_text = ", ".join(_clean_tags(self.tone_tags))
        return "\n".join(
            [
                f"type: {normalize_text(self.deed_type)}",
                f"title: {normalize_text(self.title)}",
                f"actor: {normalize_text(self.actor_name)} ({normalize_text(self.actor_id)})",
                f"depth: {int(self.depth)}",
                f"chapter: {int(self.chapter_number)}",
                f"room: {normalize_text(self.room_id)}",
                f"room_feature: {normalize_text(self.room_feature)}",
                f"tags: {tags_text}",
                f"outcome_tags: {outcome_text}",
                f"tone_tags: {tone_text}",
                f"summary: {normalize_text(self.summary)}",
            ]
        )


@dataclass
class DeedVectorizer:
    store: EmbeddingStore

    def vectorize(self, deed: Deed) -> EmbeddingRecord:
        return self.store.embed_canonical(
            source_type="deed",
            source_id=deed.deed_id,
            canonical_text=deed.canonical_text(),
        )

File: packages/engine/src/escape-the-dungeon/narrative/deeds.ts
import type { DeedMemory } from "../core/types";
import { ACTION_CONTRACTS } from "../contracts";
import {
  AnchorProjector,
  EmbeddingStore,
  HashEmbeddingProvider,
  type ProjectionBudget,
} from "../narrative/hash-embeddings";

export interface Deed {
  deedId: string;
  actorId: string;
  actorName: string;
  subjectId: string;
  sourceEntityId?: string;
  deedType: string;
  title: string;
  summary: string;
  depth: number;
  roomId: string;
  tags: string[];
  turnIndex: number;
  beliefState?: "verified" | "rumor" | "misinformed";
  confidence?: number;
}

const TRAIT_ANCHORS: Record<string, string> = {
  Comprehension: "understanding patterns and hidden causes",
  Constraint: "discipline restraint and strict control",
  Construction: "building tools practical structures engineering",
  Direction: "leadership and clear purpose",
  Empathy: "care compassion attentive listening",
  Equilibrium: "balance calm stable judgment",
  Freedom: "independence exploration improvisation",
  Levity: "humor hopeful lightness",
  Projection: "future planning and ambition",
  Survival: "resilience safety and endurance",
};

const FEATURE_ANCHORS: Record<string, string> = {
  Fame: "attention audience reach crowd engagement",
  Effort: "spent stamina sustained work exertion",
  Awareness: "perception noticing details situational awareness",
  Guile: "deception stealth opportunistic manipulation",
  Momentum: "building pace escalating progress rising pressure",
};

export class DeedVectorizer {
  private readonly store: EmbeddingStore;

  private readonly traitProjector: AnchorProjector;

  private readonly featureProjector: AnchorProjector;

  private readonly budget: ProjectionBudget;

  constructor(
    store = new EmbeddingStore(new HashEmbeddingProvider(96)),
    budget: ProjectionBudget = ACTION_CONTRACTS.deedProjection,
  ) {
    this.store = store;
    this.traitProjector = new AnchorProjector(new HashEmbeddingProvider(96), TRAIT_ANCHORS);
    this.featureProjector = new AnchorProjector(new HashEmbeddingProvider(96), FEATURE_ANCHORS);
    this.budget = budget;
  }

  canonicalText(deed: Deed): string {
    const orderedTags = [...deed.tags].sort((a, b) => a.localeCompare(b));
    return [
      `deed_id:${deed.deedId}`,
      `actor:${deed.actorId}`,
      `actor_name:${deed.actorName}`,
      `subject:${deed.subjectId}`,
      `source_entity:${deed.sourceEntityId ?? deed.actorId}`,
      `belief_state:${deed.beliefState ?? "verified"}`,
      `confidence:${Number(deed.confidence ?? 1).toFixed(2)}`,
      `type:${deed.deedType}`,
      `title:${deed.title}`,
      `summary:${deed.summary}`,
      `depth:${deed.depth}`,
      `room:${deed.roomId}`,
      `turn:${deed.turnIndex}`,
      `tags:${orderedTags.join("|")}`,
    ].join("\n");
  }

  vectorize(deed: Deed): DeedMemory {
    const canonicalText = this.canonicalText(deed);
    const record = this.store.embedCanonical("deed", deed.deedId, canonicalText);
    const traitProjection = this.traitProjector.projectVector(record.vector, 0.2, this.budget);
    const featureProjection = this.featureProjector.projectVector(record.vector, 0.2, this.budget);
    return {
      deedId: deed.deedId,
      actorEntityId: deed.actorId,
      subjectEntityId: deed.subjectId,
      summary: deed.summary,
      canonicalText,
      sourceAction: deed.deedType,
      turnIndex: deed.turnIndex,
      depth: deed.depth,
      roomId: deed.roomId,
      tags: [...deed.tags],
      beliefState: deed.beliefState ?? "verified",
      sourceEntityId: deed.sourceEntityId ?? deed.actorId,
      confidence: deed.confidence ?? 1,
      traitDelta: traitProjection.finalDeltas,
      featureDelta: featureProjection.finalDeltas,
      vector: [...record.vector],
    };
  }

  projectIntent(intentText: string): { traitDelta: Record<string, number>; featureDelta: Record<string, number> } {
    const vector = this.store.embedCanonical("intent", intentText, intentText).vector;
    const traitProjection = this.traitProjector.projectVector(vector, 0.18, this.budget);
    const featureProjection = this.featureProjector.projectVector(vector, 0.12, this.budget);
    return {
      traitDelta: traitProjection.finalDeltas,
      featureDelta: featureProjection.finalDeltas,
    };
  }

  cacheSize(): number {
    return this.store.size();
  }
}
